File: shopping-cart/shopping-cart-scala/shopping-cart/src/main/scala/com/example/shoppingcart/impl/LagomTaggerAdapter.scala
package com.example.shoppingcart.impl

import com.lightbend.lagom.scaladsl.persistence.AggregateEventTag
import com.lightbend.lagom.scaladsl.persistence.AggregateEventTagger
import com.lightbend.lagom.scaladsl.persistence.AggregateEventShards
import com.lightbend.lagom.scaladsl.persistence.AggregateEvent
import akka.cluster.sharding.typed.scaladsl.EntityContext

object LagomTaggerAdapter {

  
  def apply[Event <: AggregateEvent[Event]](
      entityCtx: EntityContext,
      lagomTagger: AggregateEventTagger[Event]
  ): Event => Set[String] = { evt =>
    val tag =
      lagomTagger match {
        case tagger: AggregateEventTag[_] =>
          tagger.tag
        case shardedTagger: AggregateEventShards[_] =>
          shardedTagger.forEntityId(entityCtx.entityId).tag
      }
    Set(tag)
  }
}

File: shopping-cart/shopping-cart-java/shopping-cart/src/main/java/com/example/shoppingcart/impl/LagomTaggerAdapter.java
package com.example.shoppingcart.impl;

import com.lightbend.lagom.javadsl.persistence.AggregateEvent;
import com.lightbend.lagom.javadsl.persistence.AggregateEventShards;
import com.lightbend.lagom.javadsl.persistence.AggregateEventTag;
import com.lightbend.lagom.javadsl.persistence.AggregateEventTagger;

import java.util.HashSet;
import java.util.Set;
import java.util.function.Function;

public class LagomTaggerAdapter {

    public static <Event extends AggregateEvent<Event>> Function<Event, Set<String>> adapt(String businessId, AggregateEventTagger<Event> lagomTagger) {
        return evt -> {
            Set<String> tags = new HashSet<>();
            if (lagomTagger instanceof AggregateEventTag) {
                tags.add(((AggregateEventTag) lagomTagger).tag());
            } else if (lagomTagger instanceof AggregateEventShards) {
                tags.add(((AggregateEventShards) lagomTagger).forEntityId(businessId).tag());
            }
            return tags;
        };

    }
}
